synapse/http/server.py

```diff
@@ -415,24 +415,59 @@ def respond_with_json(request, code, json_object, send_cors=False,
             request)
         return
 
+    encoding = "application/json"
     if pretty_print:
         json_bytes = encode_pretty_printed_json(json_object) + b"\n"
     else:
         if canonical_json or synapse.events.USE_FROZEN_DICTS:
             # canonicaljson already encodes to bytes
             json_bytes = encode_canonical_json(json_object)
         else:
-            json_bytes = json.dumps(json_object).encode("utf-8")
+            # Check what encoding the client prefers.
+            encoding = select_content_type(request)
+            logger.debug(
+                "Selected encoding %s.",
+                encoding)
+            if encoding == "application/cbor":
+                json_bytes = cbor.dumps(json_object)
+            else: # Default to json
+                logger.warn(
+                    "Unknown encoding %s, responding with application/json.",
+                    encoding)
+                encoding = "application/json"
+                json_bytes = json.dumps(json_object).encode("utf-8")
 
     return respond_with_json_bytes(
         request, code, json_bytes,
         send_cors=send_cors,
         response_code_message=response_code_message,
+        encoding=encoding
     )
 
+def select_content_type(request):
+    accepts = request.getHeader(b"Accept")
+    if accepts is None:
+        return "application/json"
+    mimetypes = accepts.decode().split(",")
+    mimetypes = [mt.strip().lower().split(";q=")for mt in mimetypes]
```

Comment:
Doesn't handle `*/*` or `sometype/*`

Reply:
This won't work on Python 3. (as an aside, do you want some help with unit tests for this code?)

Reply:
It's currently going through spec review land and this is just a demo, but yeah seeing as it's likely to be fine I'll probably like a hand with it :)